jaxlib/pytree.cc

```diff
@@ -413,6 +415,150 @@ py::object PyTreeDef::Unflatten(py::iterable leaves) const {
   }
 }
 
+py::list PyTreeDef::FlattenUpTo(py::handle xs) const {
+  py::list leaves(num_leaves());
+  std::vector<py::object> agenda;
+  agenda.push_back(py::reinterpret_borrow<py::object>(xs));
+  auto it = traversal_.rbegin();
+  int leaf = num_leaves() - 1;
+  while (!agenda.empty()) {
+    if (it == traversal_.rend()) {
+      throw std::invalid_argument(
+          "Tree structures did not match; reached end of tree.");
```

Comment:
I think a lot of these error messages could be improved by printing a shortened version of this PyTreeDef and the passed in TreeDef (and/or its structure). Let's get this in for now though.